include/Gaffer/ShufflePlug.inl

```diff
@@ -0,0 +1,121 @@
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFER_SHUFFLEPLUG_INL
+#define GAFFER_SHUFFLEPLUG_INL
+
+#include "Gaffer/Node.h"
+#include "Gaffer/PlugAlgo.h"
+
+#include <unordered_set>
+
+//////////////////////////////////////////////////////////////////////////
+// Utility methods need by ShufflesPlug::shuffle()
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+const IECore::InternedString g_sourceVariable( "source" );
+
+} // namespace
+
+namespace Gaffer
+{
+
+template<typename T>
+void ShufflesPlug::shuffle( const T &sourceContainer, T &destinationContainer ) const
+{
+	std::unordered_set<std::string> newDestinations;
+
+	const Context *context = Gaffer::Context::current();
+	Gaffer::Context::EditableScope scope( context );
+
+	for( ShufflePlugIterator it( this ); !it.done(); ++it )
+	{
+		ShufflePlug *shuffle = it->get();
+		if( !shuffle->enabledPlug()->getValue() )
+		{
+			continue;
+		}
+
+		const std::string sourcePattern = shuffle->sourcePlug()->getValue();
+		if( sourcePattern.empty() )
+		{
+			continue;
+		}
+
+		const std::string destinationPattern = shuffle->destinationPlug()->getValue();
+		if( destinationPattern.empty() )
+		{
+			continue;
+		}
+
+		bool deleteSource = shuffle->deleteSourcePlug()->getValue();
+
+		for( const auto &sourceData : sourceContainer )
+		{
+			if( !IECore::StringAlgo::match( sourceData.first, sourcePattern ) )
+			{
+				continue;
+			}
+
+			scope.set<std::string>( g_sourceVariable, sourceData.first );
```

Comment:
Its going to be a lot more `getValue()` calls if I reverse the loops... maybe I should loop once over the shuffles collecting the enabled/deleteSource/source values, and only pull destination from the inner loop?

Reply:
Addressed in 294e070

Note I did pull the relevant plug values outside the double loop. I can undo that if you prefer.

Reply:
I wasn’t positive what you meant about the dedicated hash function, but I implement my best guess in b31628c6fd76943fb1868bbec3e7bb1728f0fa73. However, my current test results don’t change with or without that commit... can you suggest an additional test idea to exercise what you were expecting to fail?

Reply:
> Note I did pull the relevant plug values outside the double loop. I can undo that if you prefer.

Without measurements, my best guess would be that it won't hurt performance, and starting by building a little data structure sets us very gently down the path towards a Shuffler, which is nice.

Reply:
> I wasn’t positive what you meant about the dedicated hash function, but I implement my best guess in b31628c. However, my current test results don’t change with or without that commit... can you suggest an additional test idea to exercise what you were expecting to fail?

I was assuming that when `destination` is computed by something that uses the `${source}` context variable, we need `${source}` to be set when `destination` gets hashed. That's really nasty because it means we need the source names at hash time. For the general case, I still think that assumption is correct. But because `Expression::hash()` deals with missing context variables, I _think_ we're OK for the anticipated use case. We should get away with an expression hash that's not varying by source name, because the source names are included in the final hash by other means (e.g. `inPlug()->attributesPlug()->hash( h )`
in `SceneElementProcessor::hashAttributes()`).

Does that make sense? To confirm we're OK, we could add a ShuffleAttributesTest which uses an expression to compute `destination`, and then run it with two sets of attributes, where the only difference between the two is the source name. If you do think I've hit upon a genuine concern (I don't think I have, but I'm somewhat fuzzy on it) then the only other option I can see is to drop the expression support for now.

Either way, we can drop  b31628c, because it'll never see `${source}` anyway without changes elsewhere.

Reply:
That makes sense. I added a test like that in ShufflePlugTest with different source CompoundObjects, but I didn't try with an actual node.

Reply:
Yeah, it needs to be on a node so that hashing/caching come in to play.

Reply:
Revert b31628c in f7cf733 rather than dropping so as not to disturb the review. Will squash that out eventually.

Added the expression test in 77f17a7